### PR TITLE
*: Use keyspan.LevelIter for rangedels in compactions

### LINKED ISSUE #1645: compaction: Use keyspan.LevelIter for rangedel iters in compactions
Currently, we front-load all rangedel blocks in all sstables participating in a compaction as part of `compaction.newInputIter`. This is inefficient and unnecessary, now that range deletions are also truncated at file bounds. We could use the new `keyspan.LevelIter` to lazily create rangedel block iterators during the course of the compaction itself, similar to how regular point key iterators are created by `levelIter`.

This optimization could be challenging if we are still subtly relying on all rangedels being exposed directly to the merging iter, however that might not be obvious until we go ahead and make this change.

### PR DESCRIPTION
**Pull request by itsbilal**
This change updates compaction.newInputIters to create
a keyspan.LevelIter containing all rangedel iterators
in a level, instead of the old way of exposing all rangedel
iterators directly to the merging iter. This makes the merging
iter more efficient as it has to manage a smaller heap (especially
for very large compactions), and also reduces the amount of
files that need to be front-opened and rangedel blocks that need
to be front-loaded before the compaction can begin. The LevelIter
can lazily open files as it goes through the compaction.

Also adds a rangedel mode to keyspan.LevelIter that makes it
find files based on point key bounds instead of on range
key bounds.

Fixes #1645

### CODE REVIEW

**Review by jbowens — COMMENTED** [review]
I think we should hold off on this until #1639 is merged, because I think there are some subtle incompatibilities. Namely, #1639 synthesizes empty spans between whole user keys. That's fine for range keys because range keys have never been written by a Pebble version that split user keys during flushes/compactions. Because the CockroachDB 22.1 migration only _marked_ the files for compaction without ensuring they're all compacted, CockroachDB 22.2 needs to still handle range deletions that may only be defined within a user key.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648#-:-N-iioti8dQrNl_7uYLy:buzz65r)* status: 0 of 4 files reviewed, all discussions resolved (waiting on @itsbilal)

**Comment by itsbilal** [comment]
Thanks for highlighting that! #1639 is merged now and while I had to tiptoe around the logic added there, this works well otherwise. I disabled the straddle span creation in rangedel mode just to keep this iterator more logically equivalent with what it replaces out of caution, but otherwise this is a pretty straightforward rebase on that change.

**Review by nicktrav — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648)* status: all files reviewed, 4 unresolved discussions (waiting on @itsbilal)

___
*[compaction.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0BsLB24aA6vMzgxG77:-N0BsLB24aA6vMzgxG78:b-j9jyl7) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/compaction.go#L1043)):*
>This is safe now that range deletions are truncated at file bounds; the merging iterator no longer needs to see all range deletes for correctness.

Just checking that there's no requirement on the _input_ side that user keys are *not* split - is this comment referring to what appears in the _output_? Concretely, I'm thinking about the Cockroach migration path: assuming this code appears in 22.2, the split user key migration might not yet have run on all tables that contain split user keys (we only mark tables as such in 22.1). If such tables appear in a compaction and hit this new codepath, I assume the split nature of the keys wont affect anything adversely?
___
*[internal/keyspan/level_iter.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N084Xm1DZrJ3foKXo-T:-N084Xm1DZrJ3foKXo-U:b-j9u07l) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L44)):*
> ```Go
> 	// we can guarantee that we will never read sstables in a level that split
> 	// user keys across them. This might be guaranteed in a future release, but
> 	// as of CockroachDB 22.2 it is no guarantee, so to be safe disable it when
> ```

nit: "it is not guaranteed"?
___
*[internal/keyspan/level_iter.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bv3kt-lhbvOxmioO_:-N0Bv3kt-lhbvOxmioOa:byjaowi) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L114)):*
> ```Go
> 	level manifest.Level,
> 	logger Logger,
> 	rangedels bool,
> ```

This is super nitpicky: I'm wondering if this could be an enum representing something like a "mode of operation" (range del mode vs. range key mode vs. combined mode), rather than a bool.

Framing that in the context of the rest of the PR, I found that whenever I encountered `if rangedels ...`, I had to mentally translate that it to "if this level iter is in range del mode, do X". Some of your comments even mention this "mode" already.

Alternatively, if you want to keep the bool, perhaps something with better readability, e.g. `if rangeDelMode`? The downside there is that there's an assumption that `!rangeDelMode` implies `rangeKeyMode`. Maybe there are cases where that's not true?
___
*[internal/keyspan/testdata/level_iter, line 445 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0ByVd43gznY6neQZSD:-N0ByVd43gznY6neQZSE:b-36g9gk) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/testdata/level_iter#L445)):*
> ```
> ok
> 
> # Test that a regular LevelIter ignores rangedels.
> ```

nit: point out that we don't produce spans for in non-range del mode?

**Review by jbowens — COMMENTED** [review]
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648)* status: all files reviewed, 8 unresolved discussions (waiting on @itsbilal)

___
*[compaction.go, line 1046 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bcpp680EvkKdOwtdS:-N0Bcpp7EW6HdzO4O6QP:b-oxkpq1) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/compaction.go#L1046)):*
> ```Go
> 		// at file bounds; the merging iterator no longer needs to see all range
> 		// deletes for correctness.
> 		wrapper := func(file *manifest.FileMetadata, iterOptions *keyspan.RangeIterOptions) (keyspan.FragmentIterator, error) {
> ```

now that the `LevelIter`, `keyspan.TableNewRangeKeyIter` and `*keyspan.RangeIterOptions` are being used for range deletions too, I think we need renaming.

maybe `keyspan.TableNewRangeKeyIter`->`keyspan.NewSpanIter`, `*keyspan.RangeIterOptions`->`*keyspan.SpanIterOptions` and rename the `Filters` field on the iter options to `RangeKeyFilters`?
___
*[internal/keyspan/level_iter.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bz9ER8MD1Rk0hzqMs:-N0Bz9ER8MD1Rk0hzqMt:b-6c4op9) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L22)):*
> ```Go
> }
> 
> // LevelIter provides a merged view of the range keys from sstables in a level.
> ```

nit: throughout this file I think we should use language that is agnostic to its use for range keys or range deletions. We can call out the two explicit uses, but we should avoid giving the reader the impression that this is an iterator exclusively for range keys.
___
*[internal/keyspan/level_iter.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0BitZIClPy8e6GTl7C:-N0BitZIClPy8e6GTl7D:b-5qjo9f) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L99)):*
> ```Go
> 	level manifest.Level,
> 	logger Logger,
> 	rangedels bool,
> ```

nit: taking `manifest.KeyType` instead of a bool would create a more readable call site. it'll be easier to understand the meaning of the argument.


edit: this is similar to nick's comment on `LevelIter.Init`.
___
*[internal/keyspan/level_iter.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0BfNxRCBhyeLHKrG-t:-N0BfNxRCBhyeLHKrG-u:b-juinqu) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L163)):*
> ```Go
> 
> 	m := l.files.SeekGE(l.cmp, key)
> 	var largestKey base.InternalKey
> ```

nit: does this local var need to be in this outer scope? looks like we could do
```
largestKey := m.LargestRangeKey
```
down below?

**Review by itsbilal — COMMENTED** [review]
TFTRs! Addressed comments but I now have to untangle the merge conflict

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648#-:-N0brRG_6aSuWBuJvfTo:b41zaxu)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @nicktrav)

___
*[compaction.go, line 1046 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bcpp680EvkKdOwtdS:-N0aTDH7-36E-bnDJmAr:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/compaction.go#L1046)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

now that the `LevelIter`, `keyspan.TableNewRangeKeyIter` and `*keyspan.RangeIterOptions` are being used for range deletions too, I think we need renaming.

maybe `keyspan.TableNewRangeKeyIter`->`keyspan.NewSpanIter`, `*keyspan.RangeIterOptions`->`*keyspan.SpanIterOptions` and rename the `Filters` field on the iter options to `RangeKeyFilters`?
</blockquote></details>

Done.
___
*[internal/keyspan/level_iter.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bz9ER8MD1Rk0hzqMs:-N0bJTKZB3HBJOEuRq9w:bcfske3) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L22)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: throughout this file I think we should use language that is agnostic to its use for range keys or range deletions. We can call out the two explicit uses, but we should avoid giving the reader the impression that this is an iterator exclusively for range keys.
</blockquote></details>

Done. Should have cleaned it all up.
___
*[internal/keyspan/level_iter.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N084Xm1DZrJ3foKXo-T:-N0bJdk25WnGJfEu5MhN:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L44)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: "it is not guaranteed"?
</blockquote></details>

Done.
___
*[internal/keyspan/level_iter.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0BitZIClPy8e6GTl7C:-N0brP8p6_9VupQQxAlX:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L99)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: taking `manifest.KeyType` instead of a bool would create a more readable call site. it'll be easier to understand the meaning of the argument.


edit: this is similar to nick's comment on `LevelIter.Init`.
</blockquote></details>

Done.
___
*[internal/keyspan/level_iter.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0Bv3kt-lhbvOxmioO_:-N0bpDYK2s7S_Vdl0Zyg:booaeui) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L114)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

This is super nitpicky: I'm wondering if this could be an enum representing something like a "mode of operation" (range del mode vs. range key mode vs. combined mode), rather than a bool.

Framing that in the context of the rest of the PR, I found that whenever I encountered `if rangedels ...`, I had to mentally translate that it to "if this level iter is in range del mode, do X". Some of your comments even mention this "mode" already.

Alternatively, if you want to keep the bool, perhaps something with better readability, e.g. `if rangeDelMode`? The downside there is that there's an assumption that `!rangeDelMode` implies `rangeKeyMode`. Maybe there are cases where that's not true?
</blockquote></details>

Done - added a KeyType like Jackson suggested above. Thanks!
___
*[internal/keyspan/level_iter.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0BfNxRCBhyeLHKrG-t:-N0bJmNH6vtmc3ybqLhD:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/level_iter.go#L163)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: does this local var need to be in this outer scope? looks like we could do
```
largestKey := m.LargestRangeKey
```
down below?
</blockquote></details>

Done.
___
*[internal/keyspan/testdata/level_iter, line 445 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0ByVd43gznY6neQZSD:-N0bJvhn5ae-HpMpchQF:b-vvu0mp) ([raw file](https://github.com/cockroachdb/pebble/blob/67dd264e6c6a43a2b4d4709b25a382f9daa05d15/internal/keyspan/testdata/level_iter#L445)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: point out that we don't produce spans for in non-range del mode?
</blockquote></details>

Not sure what you were referring to; you mean we don't produce spans for range deletions, or that we do produce synthetic/straddle spans?

**Comment by itsbilal** [comment]
Merge conflict untangled, ready for another look

**Review by jbowens — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648)* status: 0 of 7 files reviewed, 4 unresolved discussions (waiting on @itsbilal and @nicktrav)

___
*[`compaction.go` line 1070 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gxS5VAiMGvt8sjWav:-N0gxS5VAiMGvt8sjWaw:b-ic5ihi) ([raw file](https://github.com/cockroachdb/pebble/blob/b33d6e173caeeb0be7a0c26103dbc7a08defab12/compaction.go#L1070)):*
> ```Go
> 		// levelIter's purposes. While each file is still consistent before its
> 		// bounds, it's safer to have all rangedel iterators be visible to
> 		// mergingIter.
> ```

Looking at this example is making me nervous again about our handling of existing un-truncated range deletions. Would we be able to concoct a unit test testing this exact scenario? It'll probably be a little tricky given current Pebble would never create such sstable boundaries, but I think it should be possible with some elbow grease.
___
*[`internal/keyspan/level_iter.go` line 109 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gvb2e6tALq0HCicvr:-N0gvb2e6tALq0HCicvs:bd6qjv) ([raw file](https://github.com/cockroachdb/pebble/blob/45fdf3ea50cf3257a8adef3374f5561653e0386e/internal/keyspan/level_iter.go#L109)):*
> ```Go
> 	switch keyType {
> 	case manifest.KeyTypePoint:
> 		fallthrough
> ```

does `l.keyType` get set if `keyType == manifest.KeyTypePoint`? Is `KeyTypePoint` the zero value? Since this is an `Init` method, I think we should set it regardless to prevent any gotchas from `Init`-ing a `LevelIter` previously used with a different key type.
___
*[`internal/keyspan/level_iter.go` line 212 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gwSIy4AnlkCtD-Igi:-N0gwSIy4AnlkCtD-Igj:b-tnkzih) ([raw file](https://github.com/cockroachdb/pebble/blob/b33d6e173caeeb0be7a0c26103dbc7a08defab12/internal/keyspan/level_iter.go#L212)):*
> ```Go
> 
> 	// NB: the top-level Iterator has already adjusted key based on
> 	// IterOptions.UpperBound.
> ```

thanks for catching that

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648#-:-N0q78xJA-fZjuBuuAAH:b-814g8k)* status: 0 of 10 files reviewed, 4 unresolved discussions (waiting on @itsbilal, @jbowens, and @nicktrav)

___
*[`compaction.go` line 1070 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gxS5VAiMGvt8sjWav:-N0q76T7-iImM1mpkslW:bvp5klv) ([raw file](https://github.com/cockroachdb/pebble/blob/b33d6e173caeeb0be7a0c26103dbc7a08defab12/compaction.go#L1070)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Looking at this example is making me nervous again about our handling of existing un-truncated range deletions. Would we be able to concoct a unit test testing this exact scenario? It'll probably be a little tricky given current Pebble would never create such sstable boundaries, but I think it should be possible with some elbow grease.
</blockquote></details>

Done. This was a good idea.
___
*[`internal/keyspan/level_iter.go` line 109 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gvb2e6tALq0HCicvr:-N0m24Ak0oaxEJDw7Kpm:b-d0crz4) ([raw file](https://github.com/cockroachdb/pebble/blob/45fdf3ea50cf3257a8adef3374f5561653e0386e/internal/keyspan/level_iter.go#L109)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

does `l.keyType` get set if `keyType == manifest.KeyTypePoint`? Is `KeyTypePoint` the zero value? Since this is an `Init` method, I think we should set it regardless to prevent any gotchas from `Init`-ing a `LevelIter` previously used with a different key type.
</blockquote></details>

Yep it does - `fallthrough` means we run the `KeyTypeRange` case below. Go doesn't do C-style cases (fortunately? unfortunately?)

**Review by jbowens — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648#-:-N141XlA4kCCjDKQ2rxR:bnfp4nl)* status: 0 of 10 files reviewed, 3 unresolved discussions (waiting on @itsbilal and @nicktrav)

___
*[`compaction.go` line 1070 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gxS5VAiMGvt8sjWav:-N141WJrE1kWw9BYvaD4:bpfwa8y) ([raw file](https://github.com/cockroachdb/pebble/blob/b33d6e173caeeb0be7a0c26103dbc7a08defab12/compaction.go#L1070)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done. This was a good idea.
</blockquote></details>

nice, thanks!
___
*[`internal/keyspan/level_iter.go` line 109 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gvb2e6tALq0HCicvr:-N1413l40apByHCiiCep:b8t6e0f) ([raw file](https://github.com/cockroachdb/pebble/blob/45fdf3ea50cf3257a8adef3374f5561653e0386e/internal/keyspan/level_iter.go#L109)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Yep it does - `fallthrough` means we run the `KeyTypeRange` case below. Go doesn't do C-style cases (fortunately? unfortunately?)
</blockquote></details>

there is this syntax, which imo would read clearer:

```
switch keyType {
case manifest.KeyTypePoint, manifest.KeyTypeRange:
	l.keyType = keyType
	l.files = files.Filter(keyType)
default:
	panic(fmt.Sprintf("unsupported key type: %v", keyType))
}
```

**Review by itsbilal — COMMENTED** [review]
TFTR!

> Dismissed @jbowens from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1648#-:-N19GbbJ8v5unswaMpCG:b-814g8k)* status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @nicktrav)

___
*[`internal/keyspan/level_iter.go` line 109 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1648#-N0gvb2e6tALq0HCicvr:-N19GeVo5JtkhiidGJcV:b31qjmw) ([raw file](https://github.com/cockroachdb/pebble/blob/45fdf3ea50cf3257a8adef3374f5561653e0386e/internal/keyspan/level_iter.go#L109)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

there is this syntax, which imo would read clearer:

```
switch keyType {
case manifest.KeyTypePoint, manifest.KeyTypeRange:
	l.keyType = keyType
	l.files = files.Filter(keyType)
default:
	panic(fmt.Sprintf("unsupported key type: %v", keyType))
}
```
</blockquote></details>

Ah yes that's cleaner :man_facepalming: done.

**Pull request merged**
